{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import collections\n",
    "\n",
    "import nltk\n",
    "import gensim\n",
    "\n",
    "import kgglcncr.data_import as data_import\n",
    "import kgglcncr.preprocessing as preprocessing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "training_text = data_import.import_training_text()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>Cyclin-dependent kinases (CDKs) regulate a var...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>Abstract Background  Non-small cell lung cance...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>Abstract Background  Non-small cell lung cance...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>Recent evidence has demonstrated that acquired...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>Oncogenic mutations in the monomeric Casitas B...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                               Text\n",
       "0   0  Cyclin-dependent kinases (CDKs) regulate a var...\n",
       "1   1  Abstract Background  Non-small cell lung cance...\n",
       "2   2  Abstract Background  Non-small cell lung cance...\n",
       "3   3  Recent evidence has demonstrated that acquired...\n",
       "4   4  Oncogenic mutations in the monomeric Casitas B..."
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "training_text.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#training_text = preprocessing.remove_citations(training_text)\n",
    "#training_text = preprocessing.lower_case(training_text)\n",
    "\n",
    "preprocessor = preprocessing.PreprocessingPipeline([preprocessing.remove_citations,\n",
    "                                                    preprocessing.lower_case])\n",
    "\n",
    "training_text = preprocessor(training_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def bag_of_words(text_data):\n",
    "    return preprocessing.text_data_preprocessor(nltk.word_tokenize, text_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "test = bag_of_words(training_text)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>[cyclin-dependent, kinases, cdks, regulate, a,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>[abstract, background, non-small, cell, lung, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>[abstract, background, non-small, cell, lung, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>[recent, evidence, has, demonstrated, that, ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>[oncogenic, mutations, in, the, monomeric, cas...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                               Text\n",
       "0   0  [cyclin-dependent, kinases, cdks, regulate, a,...\n",
       "1   1  [abstract, background, non-small, cell, lung, ...\n",
       "2   2  [abstract, background, non-small, cell, lung, ...\n",
       "3   3  [recent, evidence, has, demonstrated, that, ac...\n",
       "4   4  [oncogenic, mutations, in, the, monomeric, cas..."
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "test.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "freq = collections.defaultdict(int)\n",
    "for i in range(len(test)):\n",
    "     for token in test.loc[i, 'Text']:\n",
    "        freq[token] += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "defaultdict(int,\n",
       "            {'nb1269': 1,\n",
       "             'hnf-1β': 7,\n",
       "             'd641g': 18,\n",
       "             'chester': 16,\n",
       "             'glu/asp': 3,\n",
       "             'venteicher': 1,\n",
       "             'sf3b1+/': 5,\n",
       "             'd5-phosphorylated': 1,\n",
       "             'nb-': 4,\n",
       "             '390–': 3,\n",
       "             'under-wound': 1,\n",
       "             'insnpg': 65,\n",
       "             'c-raf-dependence': 1,\n",
       "             'gly171asp': 4,\n",
       "             'p654l': 23,\n",
       "             'scavenging': 3,\n",
       "             '⩾100': 64,\n",
       "             'derivativeof': 1,\n",
       "             'prl-sv40': 24,\n",
       "             '93175': 1,\n",
       "             'd1-interacting': 1,\n",
       "             'lys2797argfsx26': 1,\n",
       "             'hy-line': 1,\n",
       "             'uiccstage': 1,\n",
       "             'cuny': 1,\n",
       "             'fgfr2–tacc3': 125,\n",
       "             'kex16bex11': 3,\n",
       "             'bm': 1277,\n",
       "             'glu1077': 3,\n",
       "             '-1h-pyrazolo': 2,\n",
       "             'osteosarcomas': 50,\n",
       "             'machinery—particularly': 2,\n",
       "             'self-availability': 1,\n",
       "             'notes': 31,\n",
       "             'neuhausen': 20,\n",
       "             '‘msh2': 1,\n",
       "             'c3198r': 64,\n",
       "             'foxl2lacz': 1,\n",
       "             'bhowmick': 2,\n",
       "             'paktthr308': 2,\n",
       "             \"'hits\": 2,\n",
       "             'treated/p-alk': 1,\n",
       "             'aim1l': 1,\n",
       "             'anti-snrnp70': 1,\n",
       "             '1135-bp': 1,\n",
       "             'metastasis-derived': 1,\n",
       "             'subclasses': 121,\n",
       "             'p18ink4c-deficient': 2,\n",
       "             'gonzalez-agosti': 6,\n",
       "             'inter-genic': 1,\n",
       "             'reacp53': 432,\n",
       "             'ti/tv': 1,\n",
       "             \"5'-ggg\": 2,\n",
       "             'ab26046': 2,\n",
       "             '36k': 7,\n",
       "             'l421p': 1,\n",
       "             'inhibitor”': 2,\n",
       "             'tnf-alpha': 5,\n",
       "             'ω-loop': 1,\n",
       "             'tahara': 1,\n",
       "             'played': 265,\n",
       "             'pcdna3–ddit3': 1,\n",
       "             'jsco': 1,\n",
       "             'rt−pcr': 3,\n",
       "             'c1483y': 160,\n",
       "             '160-amino-acid': 1,\n",
       "             'calypso': 20,\n",
       "             '‘5-agggacacactgggaaacag-3’': 1,\n",
       "             '7×10−5': 1,\n",
       "             'cnr': 2,\n",
       "             'six/ternary': 3,\n",
       "             'selenomethionine': 27,\n",
       "             'asbestos': 45,\n",
       "             'chemiluminiscent': 3,\n",
       "             'responsec': 9,\n",
       "             'joined': 188,\n",
       "             'r0': 80,\n",
       "             'anti-mek': 8,\n",
       "             'troubleshooting': 3,\n",
       "             'sacchromyces': 1,\n",
       "             'vector-expressed': 1,\n",
       "             'vntrs': 3,\n",
       "             '11-334': 8,\n",
       "             'spindle-checkpoint': 1,\n",
       "             '408/682': 2,\n",
       "             'telomerase-positive': 6,\n",
       "             'depending': 1580,\n",
       "             'endoscopy': 53,\n",
       "             'low–metastatic': 1,\n",
       "             'cleavage43': 3,\n",
       "             'chadwick': 4,\n",
       "             'stemke-hale': 3,\n",
       "             '18–20': 81,\n",
       "             'kiaa1033': 3,\n",
       "             'hildane': 1,\n",
       "             'thiede': 13,\n",
       "             'nony': 24,\n",
       "             'ptk7': 2,\n",
       "             'overproducer': 1,\n",
       "             'crosslink-dependent': 1,\n",
       "             'ceptor': 4,\n",
       "             'untransduced': 169,\n",
       "             '21-6': 1,\n",
       "             '94404': 1,\n",
       "             'trypsinisation': 3,\n",
       "             'charges': 352,\n",
       "             'nhp6': 2,\n",
       "             'm52': 1,\n",
       "             '~~~59yrs': 1,\n",
       "             '22/25': 1,\n",
       "             'honolulu': 2,\n",
       "             'mycn-non-amplified': 9,\n",
       "             'fus-ddit3': 14,\n",
       "             'institution’s': 4,\n",
       "             'tdt-mediated': 1,\n",
       "             'hif-2a': 2,\n",
       "             '5′-ccaggaacgtactggtgaaa-3′': 4,\n",
       "             'signified': 4,\n",
       "             '5-kilobaseecori-wi/idiii': 1,\n",
       "             'cogenics': 7,\n",
       "             'dbsnp87': 4,\n",
       "             'leave-one-out': 10,\n",
       "             'excision28': 14,\n",
       "             'androgen-containing': 7,\n",
       "             'twin': 61,\n",
       "             'oxidized': 138,\n",
       "             'gng11': 1,\n",
       "             'l-018664-00-0005': 8,\n",
       "             'whetherthe': 2,\n",
       "             '“in-house': 1,\n",
       "             'fitc-conjugated': 79,\n",
       "             '-lo-15': 1,\n",
       "             'tipped': 1,\n",
       "             'construct9': 2,\n",
       "             'cd45ro/cd45ra': 1,\n",
       "             'mmp3': 53,\n",
       "             'his231': 2,\n",
       "             '21+': 7,\n",
       "             'symmetrically': 17,\n",
       "             'pax8-pparγ': 4,\n",
       "             'm395': 23,\n",
       "             'cerulean-tagged': 5,\n",
       "             '68-amino-acid': 1,\n",
       "             'nonrearranged': 15,\n",
       "             'mac-2a': 55,\n",
       "             'k507e': 1,\n",
       "             'v600-like': 10,\n",
       "             'protein-stability': 1,\n",
       "             'raf-mitogen-activated': 2,\n",
       "             'pcdna3–myc': 2,\n",
       "             '317': 131,\n",
       "             'mis-incorporation': 3,\n",
       "             'multisizer': 1,\n",
       "             'α‐helix': 6,\n",
       "             'pt2n1m0': 4,\n",
       "             'h4-15k5d/k12d': 2,\n",
       "             'angled': 1,\n",
       "             'diode': 4,\n",
       "             'gsm180993': 1,\n",
       "             'regaining': 8,\n",
       "             'cldn18–arhgap26': 1,\n",
       "             'bellis': 2,\n",
       "             '36-base': 1,\n",
       "             'pro1073': 1,\n",
       "             'shrit1-3': 3,\n",
       "             '80-month': 1,\n",
       "             'protein–rna': 5,\n",
       "             'tsa-mediated': 4,\n",
       "             '3481': 1,\n",
       "             'y1003f': 563,\n",
       "             'third-site': 3,\n",
       "             'polyinosinic-polycytidilic': 1,\n",
       "             'genetyc': 3,\n",
       "             'dlfeir': 3,\n",
       "             'peritonitis': 2,\n",
       "             'vegfr-2': 192,\n",
       "             'immunostar': 1,\n",
       "             '4048': 3,\n",
       "             'anti-pfgfr': 1,\n",
       "             'atp/adp': 5,\n",
       "             's150-2b': 3,\n",
       "             'summarization': 16,\n",
       "             '6da': 1,\n",
       "             'g001next': 1,\n",
       "             'p53-host': 6,\n",
       "             'false-positives': 40,\n",
       "             'blanchet': 2,\n",
       "             'obviating': 20,\n",
       "             'circulatory': 22,\n",
       "             '3–14q21': 11,\n",
       "             'fallout': 4,\n",
       "             'r1699q': 1043,\n",
       "             '577hmsh6': 1,\n",
       "             'hpcs': 44,\n",
       "             '994dela': 1,\n",
       "             'fam123b/wtx': 12,\n",
       "             'com/java/colocalization/index': 8,\n",
       "             '5–1948': 2,\n",
       "             'kiaa1267': 1,\n",
       "             'v289a': 1,\n",
       "             'lumen-forming': 2,\n",
       "             'i309v': 2,\n",
       "             'beral': 2,\n",
       "             'unmanipulated': 6,\n",
       "             'pb-induced': 2,\n",
       "             '2200-6100': 1,\n",
       "             'fasay-tested': 3,\n",
       "             'generically': 4,\n",
       "             'eje-online': 7,\n",
       "             'cttgtgtcttaggctgtac': 1,\n",
       "             'a2023t': 1,\n",
       "             's3c': 338,\n",
       "             '5583': 1,\n",
       "             'conc=concentration': 13,\n",
       "             'anti-p16ink4a': 8,\n",
       "             'gln373': 1,\n",
       "             'rajaram': 1,\n",
       "             'rit': 10,\n",
       "             'ismaninger': 1,\n",
       "             'unresponsive': 99,\n",
       "             'cobrinik': 8,\n",
       "             're-sensitizing': 4,\n",
       "             'herlyn': 24,\n",
       "             'demonstrated': 10514,\n",
       "             'sp5x': 6,\n",
       "             '4650': 2,\n",
       "             'moras': 1,\n",
       "             'hpai-asuii': 3,\n",
       "             'hek-293-ddr2': 39,\n",
       "             'ubc12-14': 4,\n",
       "             '1614': 8,\n",
       "             'glu846': 6,\n",
       "             'agglomeration': 2,\n",
       "             'pdonr': 26,\n",
       "             '050-bp': 1,\n",
       "             'stereoview': 47,\n",
       "             'deficiency33': 1,\n",
       "             'e4': 62,\n",
       "             'tsr-': 2,\n",
       "             'andl187p': 1,\n",
       "             'g423a': 1,\n",
       "             '75-d-old': 1,\n",
       "             'asp450glu12': 5,\n",
       "             'cx-1/ht29': 12,\n",
       "             'aldh1a3': 1,\n",
       "             'jak2wt': 61,\n",
       "             '8-mm': 1,\n",
       "             'myxoinflammatory': 1,\n",
       "             'erbb4-kd': 7,\n",
       "             'cblr420q': 49,\n",
       "             'br41-sar73': 1,\n",
       "             'nawijn': 10,\n",
       "             'calpain': 82,\n",
       "             '5-bisphosphate–binding': 2,\n",
       "             '4768': 26,\n",
       "             'physical': 949,\n",
       "             'w5e': 2,\n",
       "             'padovan': 2,\n",
       "             'immunocytochemical': 28,\n",
       "             'trka-l92vl95v': 16,\n",
       "             'human610-quad': 17,\n",
       "             'bcl6+': 3,\n",
       "             '5′-cgg': 2,\n",
       "             'pten39-403δas': 4,\n",
       "             'tricarboxylic': 27,\n",
       "             '296g4c': 20,\n",
       "             'qukbec': 1,\n",
       "             'gene/mutation': 2,\n",
       "             'negate': 11,\n",
       "             'eschenchia': 1,\n",
       "             'cre-loxp': 10,\n",
       "             '190-base': 1,\n",
       "             'colored-coded': 26,\n",
       "             'hif-odd': 2,\n",
       "             'llcs': 5,\n",
       "             'ota': 22,\n",
       "             '1226_27delag': 1,\n",
       "             'season': 53,\n",
       "             'bix3': 1,\n",
       "             'non-asians': 7,\n",
       "             'ptp-s': 7,\n",
       "             'chloride/1': 9,\n",
       "             'ser-2': 22,\n",
       "             'zinn': 1,\n",
       "             'hip1': 64,\n",
       "             'deidentified': 8,\n",
       "             'tryprophan': 2,\n",
       "             'myc—fw': 8,\n",
       "             'originating': 402,\n",
       "             'feature': 2057,\n",
       "             '3226–3231delttaaagb': 1,\n",
       "             'thyroid': 2693,\n",
       "             'disadvantages1': 1,\n",
       "             'elf3-cooperating': 1,\n",
       "             'proteinase-k': 2,\n",
       "             'dissections': 31,\n",
       "             'hydrolytic': 41,\n",
       "             '2239t': 7,\n",
       "             '38±0': 3,\n",
       "             'purposeful': 1,\n",
       "             'csmd2': 3,\n",
       "             'a\\x82icted': 2,\n",
       "             'sticky': 1,\n",
       "             'workman': 7,\n",
       "             't9-1': 2,\n",
       "             'duodenectomy': 22,\n",
       "             '239/240': 24,\n",
       "             '1-69': 2,\n",
       "             'attracted': 22,\n",
       "             'induration': 5,\n",
       "             'trp-277': 7,\n",
       "             'directs': 159,\n",
       "             'ls270': 5,\n",
       "             '96c': 2,\n",
       "             'delays': 124,\n",
       "             'accatggatcagccagtctt': 39,\n",
       "             'r-cacctgcaagggagagaaag': 1,\n",
       "             \"'relief\": 3,\n",
       "             '50–1': 16,\n",
       "             'bfm-g': 1,\n",
       "             'vec': 262,\n",
       "             'ews/wt1–kts': 7,\n",
       "             'suzuki': 80,\n",
       "             '1239': 8,\n",
       "             'intravasate': 2,\n",
       "             'non-inflammatory': 1,\n",
       "             'ammonium-chloride-potassium': 2,\n",
       "             'hkl2map': 1,\n",
       "             'ic50s': 448,\n",
       "             'ruttenberg': 1,\n",
       "             'ap26113': 718,\n",
       "             '2d–g': 6,\n",
       "             'hifα-hydroxyproline-binding': 3,\n",
       "             '008': 244,\n",
       "             'dubs': 22,\n",
       "             'p53-mediated': 242,\n",
       "             '113-124': 12,\n",
       "             'map3k14': 2,\n",
       "             '188del11': 8,\n",
       "             'mutation-negative18': 1,\n",
       "             're-inserted': 1,\n",
       "             'berl': 1,\n",
       "             'anisopoikilocytic': 2,\n",
       "             'endured': 2,\n",
       "             \"5'-dagggggcc1aaggtcagag-3\": 1,\n",
       "             'y599f-32d-flt3': 1,\n",
       "             'cs07': 2,\n",
       "             '1181a': 1,\n",
       "             'p-alpha': 10,\n",
       "             '14a': 39,\n",
       "             'at948c': 1,\n",
       "             'abbondanza': 16,\n",
       "             'tacc1': 60,\n",
       "             '29232': 1,\n",
       "             'il-3': 3296,\n",
       "             'y1248d': 14,\n",
       "             'sept10': 3,\n",
       "             'flt3-wt/cbl': 5,\n",
       "             '6e-17': 1,\n",
       "             'afa': 11,\n",
       "             '4/289': 1,\n",
       "             'pegfp-brd4': 4,\n",
       "             '537f': 1,\n",
       "             'benefited': 164,\n",
       "             'oddity': 1,\n",
       "             'ha-psg5l': 1,\n",
       "             'd792e': 5,\n",
       "             'hindlil': 2,\n",
       "             'acidresidues': 1,\n",
       "             'dapi−': 12,\n",
       "             'joseph': 34,\n",
       "             '16q': 83,\n",
       "             'smzl': 127,\n",
       "             'mtor-responsive': 1,\n",
       "             'git2-pdgfrβ': 12,\n",
       "             'str': 169,\n",
       "             'sibaf180': 5,\n",
       "             '15/100': 2,\n",
       "             'kyse-140': 4,\n",
       "             'nhs-propionate': 1,\n",
       "             'oncogene-addicted': 39,\n",
       "             'tumours5': 1,\n",
       "             'conclusions': 1629,\n",
       "             'alã-eleshas': 1,\n",
       "             'g388r': 6,\n",
       "             'propagator': 1,\n",
       "             'transformation-resistant': 2,\n",
       "             'chromavision': 2,\n",
       "             '1050–1061': 2,\n",
       "             'g4-binding': 1,\n",
       "             'ifitm1-mediated': 2,\n",
       "             '1293–1302': 2,\n",
       "             \"'absolutely\": 2,\n",
       "             'a-p': 3,\n",
       "             'bioresearch': 1,\n",
       "             'df=2': 1,\n",
       "             '33446566': 1,\n",
       "             'p21cip1-deficient': 2,\n",
       "             'reported4': 5,\n",
       "             'development178': 1,\n",
       "             'tyr642': 14,\n",
       "             'bax-immunopositive': 1,\n",
       "             'hmlh121': 1,\n",
       "             'cbl-associated': 1,\n",
       "             'nyumc': 7,\n",
       "             'terazaki': 2,\n",
       "             '331-353': 1,\n",
       "             'a848t': 3,\n",
       "             'α-ketoglutarate1-6': 1,\n",
       "             'polyps7': 1,\n",
       "             'vose': 6,\n",
       "             '∼350': 1,\n",
       "             'crmp2': 1,\n",
       "             'kuwaitis': 1,\n",
       "             'glu-931': 6,\n",
       "             '41b': 1,\n",
       "             '1158/1541-7786': 67,\n",
       "             'p259s': 7,\n",
       "             'templeton': 5,\n",
       "             'b7-h3': 6,\n",
       "             'hyperexpression': 3,\n",
       "             'jc-hy': 12,\n",
       "             'yeastinteractions': 1,\n",
       "             'rouen': 38,\n",
       "             'progression17': 1,\n",
       "             'algorithm': 1594,\n",
       "             'easy”': 5,\n",
       "             'prc/cmv-tgf-brii': 1,\n",
       "             'neoplasticity': 1,\n",
       "             'non-disease-causing': 2,\n",
       "             '3180g20': 10,\n",
       "             'r40l': 2,\n",
       "             'mig': 58,\n",
       "             'large-cell': 173,\n",
       "             'epithelial': 4102,\n",
       "             'pik-e9or': 1,\n",
       "             'thyrotropin': 24,\n",
       "             'four-day': 1,\n",
       "             '+++/++': 3,\n",
       "             'p1-based': 1,\n",
       "             'e212∗': 4,\n",
       "             '5′-ggacagcccaagctagagatga-3′': 3,\n",
       "             'uo2509': 2,\n",
       "             'krebsforschungscentrum': 4,\n",
       "             'goswami': 1,\n",
       "             '3516-bp': 2,\n",
       "             'polymer®': 1,\n",
       "             'jacquemier': 2,\n",
       "             'impairments': 57,\n",
       "             '5′gagccctgggcactgtcgctgc3′': 1,\n",
       "             'mir-20a/b': 2,\n",
       "             'weller': 1,\n",
       "             'pi3k-transfected': 1,\n",
       "             '1756–1855': 1,\n",
       "             '2134': 13,\n",
       "             'lys22-transfected': 1,\n",
       "             'tim': 16,\n",
       "             'factor-starved': 3,\n",
       "             'diepoxybutane': 18,\n",
       "             '519k': 3,\n",
       "             'trans-splicing': 12,\n",
       "             'fr/multalin': 21,\n",
       "             'tki-targeted': 2,\n",
       "             'fhl1-induced': 1,\n",
       "             'domain–specific': 7,\n",
       "             '23398': 1,\n",
       "             'val559gly': 6,\n",
       "             'slighter': 3,\n",
       "             'chloroform/hcl': 1,\n",
       "             'scd-only': 3,\n",
       "             '2078': 3,\n",
       "             'perpetuating': 2,\n",
       "             'dehydrogenase22': 1,\n",
       "             'nfl-c-myc': 23,\n",
       "             '10-exon-11': 1,\n",
       "             'necker-enfants': 1,\n",
       "             'sscv': 12,\n",
       "             'snrk': 6,\n",
       "             'caking': 2,\n",
       "             '6/329': 2,\n",
       "             'osteoclastogenesis': 8,\n",
       "             'acctttcag': 1,\n",
       "             'nm_011113': 1,\n",
       "             'ribozymes': 3,\n",
       "             'gly499argfs*22': 1,\n",
       "             '“discovery': 5,\n",
       "             'fected': 4,\n",
       "             'magrini': 1,\n",
       "             'phoshorylation-dependent': 4,\n",
       "             'pecomas': 5,\n",
       "             'gtg': 298,\n",
       "             'ãÿ-cateninfrom': 2,\n",
       "             'contrasting': 172,\n",
       "             'anti‑cancer': 8,\n",
       "             'prelimbic': 1,\n",
       "             'bpter': 2,\n",
       "             'm483r': 14,\n",
       "             'multinucleated': 50,\n",
       "             'brdurd-negative': 1,\n",
       "             'mutation/parp': 1,\n",
       "             'hpv-59': 1,\n",
       "             'spccs': 2,\n",
       "             '743-bp': 1,\n",
       "             'vent': 5,\n",
       "             'race7': 1,\n",
       "             'lifpofectamine': 3,\n",
       "             'individual': 8506,\n",
       "             'dmrs': 2,\n",
       "             '∼44': 2,\n",
       "             'allele': 11714,\n",
       "             'met-ubiquitin': 12,\n",
       "             'aons': 30,\n",
       "             '125i-fgf2': 2,\n",
       "             'l127i': 15,\n",
       "             '11–470': 1,\n",
       "             'nm_147171': 7,\n",
       "             'calgb': 20,\n",
       "             'unblocked': 1,\n",
       "             'jak1-wt': 56,\n",
       "             'al69': 3,\n",
       "             'anti-brdu-fluos': 1,\n",
       "             'cell-cycle-dependent': 3,\n",
       "             'hemangiopericytomas': 4,\n",
       "             'psrα': 2,\n",
       "             'calabretta': 1,\n",
       "             'contrast-enhancing': 1,\n",
       "             'sult1e1': 1,\n",
       "             'pointer': 4,\n",
       "             'non‐identical': 1,\n",
       "             'slurred': 5,\n",
       "             'ttd-causing': 22,\n",
       "             'missing': 576,\n",
       "             'tp53-mutant': 7,\n",
       "             'gal4-regulated': 1,\n",
       "             'ematologia': 4,\n",
       "             '9±0': 1,\n",
       "             'non-proteolytic': 4,\n",
       "             '225-490': 2,\n",
       "             'e-predict': 2,\n",
       "             'l556_d569': 1,\n",
       "             'd04': 57,\n",
       "             'sera-lab': 1,\n",
       "             'cytogeneticist': 1,\n",
       "             'nzhangchina': 1,\n",
       "             'monoxide': 1,\n",
       "             '19/823': 1,\n",
       "             '338-1': 1,\n",
       "             'b220': 82,\n",
       "             'μg/ml/well': 1,\n",
       "             'cgcaag/gtgaga': 1,\n",
       "             'imageidentified': 2,\n",
       "             'entrapment': 2,\n",
       "             'p21/': 2,\n",
       "             'mm294': 1,\n",
       "             '0-129': 1,\n",
       "             'd1-neutralizing': 1,\n",
       "             '5-butanoate-1h-2-indolyl': 1,\n",
       "             'lmt': 12,\n",
       "             'superimposes': 22,\n",
       "             'mof': 20,\n",
       "             '−872': 1,\n",
       "             'nfi': 15,\n",
       "             'guanine‐nucleotide': 1,\n",
       "             'erbb2-mutants': 1,\n",
       "             'lesion-bound': 14,\n",
       "             'tyr608/tyr628': 2,\n",
       "             'lan-5': 4,\n",
       "             'pnp': 10,\n",
       "             'trp1837x': 2,\n",
       "             '74/178': 2,\n",
       "             '5-conjugated': 2,\n",
       "             'e18': 25,\n",
       "             'sharp58': 1,\n",
       "             'high-fat-fed': 1,\n",
       "             'mel1617': 1,\n",
       "             'mrt': 76,\n",
       "             'f/31': 3,\n",
       "             'jdp': 16,\n",
       "             'rorγt': 1,\n",
       "             'described45': 1,\n",
       "             'chr16': 11,\n",
       "             'precastrated': 8,\n",
       "             'tribolet': 1,\n",
       "             \"5'-tttgtgtgaggccg-3\": 1,\n",
       "             'autoactivity': 3,\n",
       "             'e542k-pik3ca': 1,\n",
       "             'energy-efficient': 4,\n",
       "             'f/p-negative': 5,\n",
       "             '\\xad': 1,\n",
       "             'd1673a': 1,\n",
       "             'epicardial': 8,\n",
       "             'cancer22–25': 3,\n",
       "             'gcn2': 2,\n",
       "             'hedgepeth': 27,\n",
       "             'ptp1d': 1,\n",
       "             'analysis14': 10,\n",
       "             'heiser': 2,\n",
       "             'hfgfr3': 4,\n",
       "             'wu109/wu105': 2,\n",
       "             '5557': 1,\n",
       "             'psca': 13,\n",
       "             'tyrosine-dephosphorylating': 1,\n",
       "             '5s': 6,\n",
       "             '“biological': 6,\n",
       "             'rcss46ims467': 1,\n",
       "             'ade-ura': 2,\n",
       "             'right\\\\': 2,\n",
       "             '“underexpressed': 1,\n",
       "             'unfamiliar': 2,\n",
       "             'prb-432c': 1,\n",
       "             'osada': 8,\n",
       "             'lom': 1,\n",
       "             'myelinating': 1,\n",
       "             '11–18': 77,\n",
       "             'gene-deficient': 8,\n",
       "             'loop/hook': 2,\n",
       "             'gilgfvftl': 6,\n",
       "             'seydey': 1,\n",
       "             'three-arm': 3,\n",
       "             'a-32p': 15,\n",
       "             '33p-labelled': 1,\n",
       "             '95/372': 1,\n",
       "             'fort': 9,\n",
       "             'viswanathan': 1,\n",
       "             'ser-751': 1,\n",
       "             '2712': 3,\n",
       "             'hct116p53−/−': 1,\n",
       "             'gpr18': 1,\n",
       "             'pptfig': 42,\n",
       "             'soas-2': 1,\n",
       "             'a23': 9,\n",
       "             'zake': 1,\n",
       "             'leukemia–retinoic': 1,\n",
       "             '85-': 5,\n",
       "             'ptenatgfwd': 2,\n",
       "             '81-3-5841-4778': 1,\n",
       "             'pdgfrp': 2,\n",
       "             'technitium': 3,\n",
       "             'tumours6': 14,\n",
       "             '62/837': 2,\n",
       "             'gafgs': 22,\n",
       "             'egfr-g719a+s768i': 5,\n",
       "             'nontransmissions': 2,\n",
       "             '2-micron': 1,\n",
       "             'phosphorylates': 837,\n",
       "             'near-activation': 4,\n",
       "             'n564d': 63,\n",
       "             'alkr1275': 13,\n",
       "             'fig3b': 5,\n",
       "             'l833fa': 1,\n",
       "             'h-1205-011-4': 4,\n",
       "             'wm1361': 29,\n",
       "             't2n0m0': 8,\n",
       "             'er/pr/her2': 9,\n",
       "             'uas-alkf1174l': 7,\n",
       "             't-cell-independent': 1,\n",
       "             'thinner': 7,\n",
       "             'tweak-receptor': 1,\n",
       "             'preanalytix': 6,\n",
       "             's79': 2,\n",
       "             'unradiolabeled': 1,\n",
       "             'nct00113763': 1,\n",
       "             'muttttttgagagtgc': 1,\n",
       "             'nflg': 1,\n",
       "             'pmscv-eyfp-ires': 1,\n",
       "             'ascom-mll4': 9,\n",
       "             'pcm1': 76,\n",
       "             'odero': 1,\n",
       "             'silico_': 13,\n",
       "             'processes/pathways': 7,\n",
       "             'adenosquamous-carcinomas': 1,\n",
       "             'nm_016931': 1,\n",
       "             'anti-glut1': 13,\n",
       "             'preincubation': 38,\n",
       "             'acaca': 2,\n",
       "             'tric-specific': 28,\n",
       "             'bollo': 2,\n",
       "             'del703-37': 1,\n",
       "             '1217': 4,\n",
       "             'ghmsh2': 1,\n",
       "             'nct01476137': 1,\n",
       "             '93-i-0063': 2,\n",
       "             '9myc': 1,\n",
       "             'valine': 695,\n",
       "             'cancers8': 3,\n",
       "             'brugarolas': 15,\n",
       "             'radioinert': 5,\n",
       "             'unexpected': 802,\n",
       "             'a_52_p70381': 1,\n",
       "             '20mb': 5,\n",
       "             'proband': 1298,\n",
       "             'pnfs18': 1,\n",
       "             'anti–t-cell': 1,\n",
       "             'casp5': 1,\n",
       "             '6830': 1,\n",
       "             '374-385': 1,\n",
       "             'overtmorphological': 1,\n",
       "             'amag': 8,\n",
       "             'predications': 13,\n",
       "             'homoz': 4,\n",
       "             'cre-recombination': 4,\n",
       "             'end-points': 16,\n",
       "             'abrogating': 145,\n",
       "             'cgh': 827,\n",
       "             'sarms': 6,\n",
       "             'deenqq': 1,\n",
       "             'aatgtccatctgataagacg': 1,\n",
       "             'wagr': 17,\n",
       "             '‘exemplars’': 1,\n",
       "             'flt3-d835y': 21,\n",
       "             'svegfr2': 14,\n",
       "             'claudin': 13,\n",
       "             'mariani': 5,\n",
       "             'arvo': 3,\n",
       "             'composing': 11,\n",
       "             '5’-cttgacaatgtgtacg': 3,\n",
       "             'map3k12': 2,\n",
       "             'scancontrol': 1,\n",
       "             '4987c→g': 64,\n",
       "             'trpthr': 1,\n",
       "             'jak-homology': 1,\n",
       "             'itim': 3,\n",
       "             'pathophysiologically': 5,\n",
       "             'irides': 3,\n",
       "             'polydramnios': 5,\n",
       "             'tumorigenic16': 1,\n",
       "             'ariz': 2,\n",
       "             'hr=19': 1,\n",
       "             'akt2-r170w': 9,\n",
       "             'chloride–induced': 1,\n",
       "             'i36g': 1,\n",
       "             'tgr-1': 1,\n",
       "             'proteasomal-mediated': 1,\n",
       "             'a1708er': 1,\n",
       "             'bt325': 15,\n",
       "             'egf-r': 9,\n",
       "             'e2-24k': 2,\n",
       "             'cycle': 8944,\n",
       "             '237–269': 2,\n",
       "             'zbtb46': 1,\n",
       "             '2114insa': 2,\n",
       "             'audiological': 4,\n",
       "             'delivery8': 5,\n",
       "             'bicc1-fgfr2': 5,\n",
       "             'veneto': 8,\n",
       "             'achong': 4,\n",
       "             'tae68414': 1,\n",
       "             'associationwith': 2,\n",
       "             '2×5': 2,\n",
       "             'cnag': 8,\n",
       "             '29months': 1,\n",
       "             '94e-53': 1,\n",
       "             '5-cgggatccgccaccatggacttcagcagaaatc-3': 1,\n",
       "             'hrasq61l-mediated': 17,\n",
       "             'kg/ml': 1,\n",
       "             'asp12': 2,\n",
       "             'tissue44': 7,\n",
       "             'trypsinized': 367,\n",
       "             'transit293': 1,\n",
       "             'icig-7h': 1,\n",
       "             'alpha-carboxylate': 4,\n",
       "             'fgf16–fgf23': 5,\n",
       "             '4eb-p1': 1,\n",
       "             'aqp3': 2,\n",
       "             'kr': 67,\n",
       "             'peaking': 9,\n",
       "             '11q14-qter': 1,\n",
       "             'alsoact': 1,\n",
       "             'window”': 21,\n",
       "             'desirable': 109,\n",
       "             'fissure': 5,\n",
       "             'xwx': 2,\n",
       "             'phosphatase–conjugated': 7,\n",
       "             'aeolicus': 20,\n",
       "             '5022': 7,\n",
       "             '40-day-old': 1,\n",
       "             'two-syringe': 1,\n",
       "             'custom-made': 21,\n",
       "             'aaggcgggcccggga': 1,\n",
       "             '8/84': 5,\n",
       "             '4±4': 3,\n",
       "             'nathaniel': 1,\n",
       "             'fluvastatin': 4,\n",
       "             '~52': 3,\n",
       "             '13·5': 13,\n",
       "             'pu6': 1,\n",
       "             'pegfp-δc': 1,\n",
       "             'supposed': 104,\n",
       "             'rawat': 2,\n",
       "             'vemuraftenib': 4,\n",
       "             'multi-label': 2,\n",
       "             '5′-cagcgatgcgtacacaga-3′': 1,\n",
       "             'pathania': 1,\n",
       "             'a859t': 82,\n",
       "             'd11s913': 1,\n",
       "             'epd24': 10,\n",
       "             'aml-mrc': 2,\n",
       "             'culture-insert': 1,\n",
       "             'flt-1': 13,\n",
       "             '1n8z': 16,\n",
       "             '29/39': 10,\n",
       "             'genotype-phenotypecorrelations': 2,\n",
       "             'gel-shift': 38,\n",
       "             'th1': 19,\n",
       "             'furiosus7': 1,\n",
       "             '551101': 3,\n",
       "             'n2=13': 2,\n",
       "             'ofnup98': 6,\n",
       "             'k218t': 30,\n",
       "             'hccc-type': 3,\n",
       "             '0/38': 1,\n",
       "             'cholinergic': 20,\n",
       "             'd–h•••a': 5,\n",
       "             'akaps': 2,\n",
       "             'ab32072': 2,\n",
       "             'hypoacetylation': 29,\n",
       "             'speculations': 18,\n",
       "             'juxtamenbrane': 1,\n",
       "             'jab1317': 1,\n",
       "             'institute-frederick': 3,\n",
       "             'g779c': 1,\n",
       "             'sum149': 60,\n",
       "             '11-fold': 29,\n",
       "             'serine-proline': 10,\n",
       "             '33-37': 3,\n",
       "             't-all17': 1,\n",
       "             'pjp11': 3,\n",
       "             'enthusiasm': 24,\n",
       "             'hre': 12,\n",
       "             'macro-complexes': 1,\n",
       "             'cys47phe': 8,\n",
       "             'co2+-nta': 1,\n",
       "             '1835c': 1,\n",
       "             'cag→ctg': 3,\n",
       "             '35/58': 2,\n",
       "             'astra': 12,\n",
       "             '131–310': 1,\n",
       "             '117599': 1,\n",
       "             'f264a': 1,\n",
       "             '1-flag': 2,\n",
       "             'apc580s': 1,\n",
       "             '1p': 308,\n",
       "             'x-100/0': 9,\n",
       "             's1655': 37,\n",
       "             '9α-fluorocortisol': 3,\n",
       "             'rad51l2': 3,\n",
       "             '7/66': 5,\n",
       "             'anti-cd95': 8,\n",
       "             'decidualization': 4,\n",
       "             'davidson': 5,\n",
       "             'scalepack': 31,\n",
       "             'growth4': 1,\n",
       "             'naphthol-as-biphosphate': 1,\n",
       "             'tetrachloride': 1,\n",
       "             'leucine-rich': 55,\n",
       "             'p170': 5,\n",
       "             'ttcacacacaaac': 12,\n",
       "             'antiphosphothreonine': 7,\n",
       "             'allocates': 1,\n",
       "             '0002158': 20,\n",
       "             'shuin': 1,\n",
       "             'cfu-gemms': 1,\n",
       "             'chains': 1673,\n",
       "             's566': 8,\n",
       "             'model95': 1,\n",
       "             'a272': 9,\n",
       "             'miliary': 12,\n",
       "             '6–17': 3,\n",
       "             '520-': 12,\n",
       "             'asp820gly': 5,\n",
       "             'tc-3': 4,\n",
       "             'localisations': 2,\n",
       "             'k52r': 94,\n",
       "             'lam': 65,\n",
       "             'pik3ca-forward': 1,\n",
       "             '3478': 1,\n",
       "             'aminoactinomycin': 1,\n",
       "             'favouring': 36,\n",
       "             'gpi-linked': 2,\n",
       "             'brentuximab': 8,\n",
       "             '3972': 4,\n",
       "             'sur8rf': 1,\n",
       "             '4–147': 1,\n",
       "             'k-ras-g12c': 33,\n",
       "             '105f10': 1,\n",
       "             'sh2': 1354,\n",
       "             '1000-3000': 2,\n",
       "             'identical-sized': 5,\n",
       "             'pmf–mutated': 1,\n",
       "             'manipulations': 71,\n",
       "             'v677a': 1,\n",
       "             'tbp-associated': 6,\n",
       "             'mtsl': 5,\n",
       "             '1087–1093': 1,\n",
       "             'tov774': 1,\n",
       "             'receiver-impairing': 4,\n",
       "             '1039-2329_1409+827del': 1,\n",
       "             'gentien': 14,\n",
       "             'ab5443': 6,\n",
       "             'membrane-association': 1,\n",
       "             'scienti': 2,\n",
       "             'taufkirchen': 10,\n",
       "             '125-kda': 4,\n",
       "             'volocity': 7,\n",
       "             'flag-type': 36,\n",
       "             'gst-box': 14,\n",
       "             'minvielle': 1,\n",
       "             'moderated': 13,\n",
       "             '5d5d': 8,\n",
       "             'phrap': 7,\n",
       "             'food': 348,\n",
       "             'pc53-sn3-t123a': 3,\n",
       "             '1–302': 27,\n",
       "             'myoblasts': 120,\n",
       "             'spop-w131g': 14,\n",
       "             'â€œtranscriptinstabilityâ€•are': 1,\n",
       "             'ptrip/δu3–ef1-α–egfp': 2,\n",
       "             'ladeb': 2,\n",
       "             'mosaicism’': 1,\n",
       "             'ossareh-nazari': 1,\n",
       "             'l16d3': 1,\n",
       "             'conclusions—the': 1,\n",
       "             'brd4-imposed': 1,\n",
       "             '114179': 1,\n",
       "             'ecm-receptor': 4,\n",
       "             'y139d': 99,\n",
       "             'antihost': 2,\n",
       "             'high-': 39,\n",
       "             'h514r': 1,\n",
       "             't341k': 2,\n",
       "             'unicz': 1,\n",
       "             'sbalk': 2,\n",
       "             '71-gene': 1,\n",
       "             'h1421y-l2': 12,\n",
       "             'plots': 730,\n",
       "             'uf2': 6,\n",
       "             'brafis': 36,\n",
       "             'immunology': 37,\n",
       "             'mutp53': 33,\n",
       "             'nuleotides': 1,\n",
       "             'y98c': 2,\n",
       "             'p19arf-deficient': 1,\n",
       "             'δ14–15': 4,\n",
       "             'pbabe-puror-flag-snf5': 1,\n",
       "             'represses': 317,\n",
       "             'zc3h7b': 26,\n",
       "             'stillman': 3,\n",
       "             '39-years-old': 1,\n",
       "             '21-25': 10,\n",
       "             \"'classic\": 2,\n",
       "             '5´10-9': 2,\n",
       "             '1a–i': 1,\n",
       "             'raf-1-14-3-3': 1,\n",
       "             '200-μl': 7,\n",
       "             'cycle-type': 1,\n",
       "             '14/26': 2,\n",
       "             'd16s3095': 4,\n",
       "             'ad-hies': 39,\n",
       "             'studies5–7': 5,\n",
       "             'psmads': 1,\n",
       "             'mea': 1,\n",
       "             'lys753': 1,\n",
       "             '1-500': 1,\n",
       "             'e138': 2,\n",
       "             'deprotected': 1,\n",
       "             'r175f': 1,\n",
       "             'anti-keratin': 1,\n",
       "             'y204': 2,\n",
       "             'gst-vhl-elonginc-elonginb': 18,\n",
       "             'c290r': 6,\n",
       "             'series”': 1,\n",
       "             'asp537': 14,\n",
       "             '3–44': 2,\n",
       "             'p-mek': 85,\n",
       "             'fuscc': 7,\n",
       "             'm774': 4,\n",
       "             'cellularity16': 5,\n",
       "             '31i': 1,\n",
       "             'dnareplication': 1,\n",
       "             'sinnamon': 1,\n",
       "             '6/2014': 1,\n",
       "             'r743pfsx13': 1,\n",
       "             'mmc-inducible': 1,\n",
       "             'chemiluminescent': 197,\n",
       "             'rp11': 4,\n",
       "             'tu-02': 2,\n",
       "             'transductional': 1,\n",
       "             'erythroblastosis': 33,\n",
       "             'hxxlαs': 1,\n",
       "             'pen–strep': 1,\n",
       "             'laci-cherry': 11,\n",
       "             't3/∆leu': 6,\n",
       "             '5′-tgtggtttcactttaaaccctgga-3′': 5,\n",
       "             '“boost/nest”': 1,\n",
       "             'vicinity': 441,\n",
       "             '1dfigure': 1,\n",
       "             '28046': 2,\n",
       "             'termuhlen': 3,\n",
       "             'mnsg': 2,\n",
       "             'lineage': 1304,\n",
       "             ...})"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "freq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "248762"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(freq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def _word_frequency_cutoff(bag_of_words):\n",
    "    return [token for token in bag_of_words if freq[token] > 1]\n",
    "def word_frequency_cutoff(text_data):\n",
    "    return preprocessing.text_data_preprocessor(_word_frequency_cutoff, text_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "bow = word_frequency_cutoff(test)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>[cyclin-dependent, kinases, cdks, regulate, a,...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>[abstract, background, non-small, cell, lung, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>[abstract, background, non-small, cell, lung, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>[recent, evidence, has, demonstrated, that, ac...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>[oncogenic, mutations, in, the, monomeric, cas...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                               Text\n",
       "0   0  [cyclin-dependent, kinases, cdks, regulate, a,...\n",
       "1   1  [abstract, background, non-small, cell, lung, ...\n",
       "2   2  [abstract, background, non-small, cell, lung, ...\n",
       "3   3  [recent, evidence, has, demonstrated, that, ac...\n",
       "4   4  [oncogenic, mutations, in, the, monomeric, cas..."
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bow.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "dictionary = gensim.corpora.Dictionary(bow.loc[:,'Text'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Dictionary(157719 unique tokens: ['us/intergroup', 'd38n', 'fashion18', 'hnf-1β', 'd641g']...)\n"
     ]
    }
   ],
   "source": [
    "print(dictionary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def count_feature(text_data, dictionary):\n",
    "    return preprocessing.text_data_preprocessor(dictionary.doc2bow, text_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "fv = count_feature(bow, dictionary)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Text</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>[(0, 2), (1, 3), (2, 14), (3, 1), (4, 2), (5, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>[(0, 7), (2, 1), (4, 1), (8, 1), (9, 2), (11, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>[(0, 7), (2, 1), (4, 1), (8, 1), (9, 2), (11, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>[(0, 1), (2, 4), (5, 1), (6, 1), (8, 1), (11, ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>[(0, 4), (2, 6), (6, 1), (9, 1), (11, 336), (1...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ID                                               Text\n",
       "0   0  [(0, 2), (1, 3), (2, 14), (3, 1), (4, 2), (5, ...\n",
       "1   1  [(0, 7), (2, 1), (4, 1), (8, 1), (9, 2), (11, ...\n",
       "2   2  [(0, 7), (2, 1), (4, 1), (8, 1), (9, 2), (11, ...\n",
       "3   3  [(0, 1), (2, 4), (5, 1), (6, 1), (8, 1), (11, ...\n",
       "4   4  [(0, 4), (2, 6), (6, 1), (9, 1), (11, 336), (1..."
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "tfidf = gensim.models.TfidfModel(fv.loc[:,'Text'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "tfidf = tfidf[fv.loc[:,'Text']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<gensim.interfaces.TransformedCorpus at 0x7ff2797c1978>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tfidf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "lsi = gensim.models.LsiModel(tfidf, id2word=dictionary, num_topics=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "corpus_lsi = lsi[tfidf]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.654*\"brca1\" + 0.306*\"brct\" + 0.258*\"vus\" + 0.200*\"variants\" + 0.129*\"neutral\" + 0.123*\"⇓\" + 0.119*\"deleterious\" + 0.118*\"nih-pa\" + 0.114*\"brca2\" + 0.110*\"−2\" + 0.101*\"manuscript\" + 0.089*\"classifi\" + 0.083*\"author\" + 0.072*\"−1\" + 0.071*\"phosphopeptide\" + 0.068*\"brca\" + 0.067*\"e2\" + 0.066*\"m1775r\" + 0.063*\"variant\" + 0.059*\"transcriptional\" + 0.058*\"vuss\" + 0.057*\"−3\" + 0.055*\"bard1\" + 0.055*\"rmce\" + 0.055*\"ed\" + 0.054*\"odds\" + 0.053*\"117\" + 0.052*\"cisplatin\" + 0.052*\"classification\" + 0.045*\"risk\" + 0.044*\"cation\" + 0.043*\"iarc\" + 0.042*\"cient\" + 0.042*\"proteolysis\" + 0.042*\"ovarian\" + 0.041*\"breast\" + 0.040*\"pathogenic\" + 0.040*\"pmc\" + 0.040*\"causality\" + 0.039*\"hdr\" + 0.039*\"yeast\" + 0.039*\"supplementary\" + 0.039*\"ssa\" + 0.037*\"hr\" + 0.036*\"pten\" + 0.036*\"missense\" + 0.036*\"loh\" + 0.035*\"stem\" + 0.034*\"neighborhood\" + 0.034*\"−\" + 0.033*\"peptide\" + 0.033*\">\" + 0.033*\"classified\" + 0.032*\"unclassified\" + 0.031*\"history\" + 0.031*\"res\" + 0.031*\"embryonic\" + 0.031*\"a1708e\" + 0.031*\"ring\" + 0.030*\"fig\" + 0.030*\"class\" + 0.029*\"repair\" + 0.029*\"v1736a\" + 0.029*\"june\" + 0.029*\"likelihood\" + 0.029*\"complementation\" + 0.029*\"2013\" + 0.028*\"assays\" + 0.028*\"scp\" + 0.028*\"m1689r\" + 0.028*\"p53\" + 0.028*\"aacrjournals\" + 0.028*\"align-gvgd\" + 0.028*\"+3\" + 0.028*\"stability\" + 0.028*\"ubiquitin\" + 0.027*\"buried\" + 0.027*\"cd-13-0094\" + 0.027*\"r1699q\" + 0.026*\"r1699w\" + 0.026*\"folding\" + 0.026*\"effect”\" + 0.026*\"1158/2159-8290\" + 0.026*\"egfr\" + 0.026*\"m1775k\" + 0.026*\"repeats\" + 0.026*\"ligase\" + 0.025*\"uncertain\" + 0.025*\"individuals\" + 0.025*\"cancer-associated\" + 0.025*\"recombination\" + 0.025*\"pathogenicity\" + 0.025*\"trypsin\" + 0.025*\"c0\" + 0.024*\"gv-gd\" + 0.024*\"splicing\" + 0.024*\"c65\" + 0.024*\"limits\" + 0.024*\"cancerdiscovery\" + 0.023*\"segregation\"'"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lsi.print_topic(0,100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'-0.270*\"epas1\" + -0.233*\"jak2\" + 0.211*\"jak1\" + -0.206*\"hif2α\" + 0.196*\"sox9\" + -0.172*\"mkk4\" + -0.140*\"elf3\" + 0.126*\"tric\" + -0.119*\"crenolanib\" + 0.109*\"ezh2\"'"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lsi.print_topic(99)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "lsi2 =  gensim.models.LsiModel(tfidf, id2word=dictionary, num_topics=10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(0,\n",
       "  '-0.654*\"brca1\" + -0.306*\"brct\" + -0.258*\"vus\" + -0.200*\"variants\" + -0.129*\"neutral\" + -0.123*\"⇓\" + -0.119*\"deleterious\" + -0.118*\"nih-pa\" + -0.114*\"brca2\" + -0.110*\"−2\" + -0.101*\"manuscript\" + -0.089*\"classifi\" + -0.083*\"author\" + -0.072*\"−1\" + -0.071*\"phosphopeptide\" + -0.068*\"brca\" + -0.067*\"e2\" + -0.066*\"m1775r\" + -0.063*\"variant\" + -0.059*\"transcriptional\" + -0.058*\"vuss\" + -0.057*\"−3\" + -0.055*\"bard1\" + -0.055*\"rmce\" + -0.055*\"ed\" + -0.054*\"odds\" + -0.053*\"117\" + -0.052*\"cisplatin\" + -0.052*\"classification\" + -0.045*\"risk\"'),\n",
       " (1,\n",
       "  '-0.778*\"pten\" + -0.208*\"phts\" + -0.184*\"pip3\" + -0.150*\"phosphatase\" + -0.150*\"asd/dd\" + -0.100*\"asp92\" + -0.092*\"tumor-related\" + -0.086*\"egfr\" + -0.085*\"catalytic\" + 0.077*\"brca1\" + -0.076*\"yeast\" + -0.065*\"ptdins\" + -0.064*\"p-loop\" + -0.063*\"p110α-caax\" + -0.063*\"pten-like\" + -0.061*\"p3\" + -0.059*\"p4\" + -0.056*\"loops\" + -0.050*\"6-pten\" + -0.050*\"h93r\" + -0.050*\"ins\" + -0.048*\"gefitinib\" + -0.047*\"lipid\" + -0.047*\"pi3k\" + -0.046*\"autism\" + -0.045*\"gfp-akt1\" + -0.044*\"ptens\" + -0.044*\"asd/dd-related\" + -0.044*\"tpip\" + -0.044*\"wpd-\"'),\n",
       " (2,\n",
       "  '-0.570*\"egfr\" + -0.333*\"gefitinib\" + 0.187*\"pten\" + -0.146*\"erlotinib\" + -0.137*\"l858r\" + -0.130*\"imatinib\" + -0.128*\"nsclc\" + -0.119*\"alk\" + 0.106*\"brca1\" + -0.103*\"pdgfra\" + -0.100*\"exon\" + -0.091*\"tkis\" + -0.076*\"erbb2\" + -0.073*\"resistance\" + -0.072*\"tki\" + -0.069*\"lung\" + -0.069*\"braf\" + -0.065*\"gists\" + -0.065*\"patients\" + -0.064*\"gist\" + -0.062*\"ras\" + -0.062*\"ba/f3\" + -0.061*\"months\" + -0.059*\"kit\" + -0.058*\"her2\" + -0.058*\"kinase\" + -0.058*\"pfs\" + 0.056*\"phts\" + -0.056*\"t790m\" + -0.055*\"lapatinib\"'),\n",
       " (3,\n",
       "  '0.490*\"−2\" + 0.320*\"−1\" + -0.277*\"brct\" + 0.260*\"vus\" + 0.253*\"−3\" + 0.223*\"brca2\" + 0.178*\"deleterious\" + 0.171*\"brca\" + 0.142*\"−\" + -0.139*\"⇓\" + 0.136*\"odds\" + 0.124*\"vuss\" + -0.115*\"brca1\" + -0.103*\"nih-pa\" + 0.100*\"−4\" + -0.089*\"manuscript\" + 0.080*\"causality\" + -0.080*\"phosphopeptide\" + 0.076*\"00\" + -0.071*\"author\" + -0.069*\"m1775r\" + -0.064*\"transcriptional\" + 0.063*\"neutral\" + 0.056*\"−5\" + -0.055*\"117\" + 0.054*\"favor\" + 0.053*\"loh\" + 0.051*\"433\" + -0.048*\"proteolysis\" + 0.046*\"myriad\"'),\n",
       " (4,\n",
       "  '-0.395*\"egfr\" + -0.250*\"gefitinib\" + 0.210*\"imatinib\" + 0.205*\"pdgfra\" + 0.190*\"alk\" + -0.163*\"vus\" + -0.162*\"brca1\" + 0.155*\"tsc2\" + 0.131*\"p53\" + 0.118*\"gists\" + 0.112*\"gist\" + 0.111*\"−2\" + -0.106*\"erlotinib\" + -0.105*\"l858r\" + 0.102*\"mtor\" + 0.101*\"ras\" + 0.095*\"braf\" + 0.093*\"tsc1\" + 0.088*\"brct\" + 0.088*\"kit\" + -0.088*\"classifi\" + 0.080*\"variants\" + -0.077*\"pten\" + 0.074*\"−1\" + 0.071*\"crizotinib\" + 0.068*\"melanoma\" + 0.066*\"crenolanib\" + 0.065*\"flt3\" + 0.064*\"b-raf\" + -0.061*\"neutral\"'),\n",
       " (5,\n",
       "  '0.337*\"vus\" + -0.310*\"brct\" + -0.294*\"−2\" + -0.192*\"−1\" + 0.164*\"classifi\" + 0.160*\"imatinib\" + 0.156*\"pdgfra\" + 0.155*\"brca1\" + -0.154*\"nih-pa\" + -0.152*\"−3\" + -0.149*\"egfr\" + -0.138*\"⇓\" + -0.126*\"manuscript\" + 0.120*\"neutral\" + 0.101*\"rmce\" + 0.101*\"ed\" + -0.098*\"author\" + -0.096*\"gefitinib\" + -0.096*\"phosphopeptide\" + 0.096*\"alk\" + 0.095*\"cisplatin\" + -0.094*\"variants\" + 0.091*\"gists\" + 0.085*\"gist\" + -0.085*\"−\" + 0.083*\"cation\" + 0.079*\"cient\" + -0.075*\"117\" + -0.074*\"m1775r\" + 0.067*\"stem\"'),\n",
       " (6,\n",
       "  '-0.643*\"tsc2\" + -0.375*\"tsc1\" + 0.211*\"imatinib\" + 0.210*\"pdgfra\" + -0.170*\"s6k\" + -0.166*\"mtor\" + -0.157*\"tsc\" + -0.153*\"t389/s6k\" + -0.145*\"tsc1–tsc2\" + 0.123*\"gists\" + 0.114*\"gist\" + 0.081*\"kit\" + -0.075*\"r611q\" + -0.074*\"mtorc1\" + 0.071*\"crenolanib\" + -0.070*\"variants\" + -0.065*\"supp\" + -0.065*\"pathogenic\" + -0.065*\"torc1\" + 0.064*\"−2\" + 0.059*\"brct\" + 0.058*\"sunitinib\" + -0.055*\"sift\" + -0.050*\"variant\" + -0.050*\"tuberin\" + 0.049*\"ponatinib\" + -0.048*\"q1554h\" + -0.047*\"egfr\" + -0.046*\"vus\" + 0.046*\"d842v\"'),\n",
       " (7,\n",
       "  '-0.696*\"alk\" + 0.256*\"pdgfra\" + -0.256*\"crizotinib\" + 0.250*\"imatinib\" + 0.168*\"tsc2\" + -0.155*\"neuroblastoma\" + 0.150*\"gists\" + 0.138*\"gist\" + 0.098*\"kit\" + 0.096*\"tsc1\" + 0.087*\"crenolanib\" + -0.086*\"ceritinib\" + -0.075*\"ros1\" + -0.072*\"tkd\" + 0.070*\"sunitinib\" + -0.068*\"f1174l\" + 0.060*\"exon\" + 0.056*\"d842v\" + -0.056*\"l1196m\" + 0.056*\"ponatinib\" + -0.054*\"pf-06463922\" + -0.053*\"eml4-alk\" + -0.046*\"ap26113\" + -0.044*\"asp3026\" + -0.044*\"h3122\" + 0.044*\"v654a\" + 0.044*\"s6k\" + 0.042*\"dod\" + -0.042*\"mycn\" + 0.041*\"t389/s6k\"'),\n",
       " (8,\n",
       "  '-0.374*\"alk\" + 0.281*\"p53\" + -0.270*\"tsc2\" + -0.201*\"pdgfra\" + -0.197*\"imatinib\" + 0.167*\"smad2\" + 0.166*\"smad4\" + 0.157*\"ras\" + -0.151*\"tsc1\" + -0.140*\"crizotinib\" + 0.128*\"smad3\" + 0.126*\"mtor\" + -0.119*\"gists\" + -0.109*\"gist\" + 0.107*\"braf\" + 0.106*\"b-raf\" + 0.098*\"p53s\" + -0.079*\"neuroblastoma\" + 0.079*\"kras\" + 0.078*\"notch\" + -0.073*\"kit\" + -0.069*\"crenolanib\" + -0.068*\"s6k\" + 0.068*\"pik3ca\" + -0.067*\"t389/s6k\" + -0.067*\"brct\" + 0.066*\"hotspots\" + -0.063*\"tsc1–tsc2\" + -0.062*\"tsc\" + 0.062*\"c-raf\"'),\n",
       " (9,\n",
       "  '0.463*\"mtor\" + -0.439*\"smad2\" + -0.430*\"smad4\" + -0.333*\"smad3\" + 0.202*\"mtorc1\" + -0.126*\"mh2\" + -0.100*\"tsc2\" + -0.095*\"smad\" + 0.092*\"s2215y\" + 0.090*\"rapamycin\" + -0.077*\"tgf-β\" + 0.074*\"deptor\" + 0.071*\"mtorc2\" + -0.064*\"heterocomplex\" + -0.062*\"homotrimer\" + -0.058*\"r-smads\" + 0.058*\"rheb\" + -0.057*\"mh1\" + 0.054*\"s6k1\" + 0.052*\"r2505p\" + 0.051*\"ras\" + -0.051*\"r-smad\" + -0.050*\"sara\" + -0.050*\"tsc1\" + 0.048*\"nutrient\" + -0.048*\"smad2-smad4\" + -0.045*\"alk\" + -0.045*\"mmr\" + -0.043*\"crcs\" + 0.043*\"pik3ca\"')]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lsi2.print_topics(num_words=30)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on class LsiModel in module gensim.models.lsimodel:\n",
      "\n",
      "class LsiModel(gensim.interfaces.TransformationABC, gensim.models.basemodel.BaseTopicModel)\n",
      " |  Objects of this class allow building and maintaining a model for Latent\n",
      " |  Semantic Indexing (also known as Latent Semantic Analysis).\n",
      " |  \n",
      " |  The main methods are:\n",
      " |  \n",
      " |  1. constructor, which initializes the projection into latent topics space,\n",
      " |  2. the ``[]`` method, which returns representation of any input document in the\n",
      " |     latent space,\n",
      " |  3. `add_documents()` for incrementally updating the model with new documents.\n",
      " |  \n",
      " |  The left singular vectors are stored in `lsi.projection.u`, singular values\n",
      " |  in `lsi.projection.s`. Right singular vectors can be reconstructed from the output\n",
      " |  of `lsi[training_corpus]`, if needed. See also FAQ [2]_.\n",
      " |  \n",
      " |  Model persistency is achieved via its load/save methods.\n",
      " |  \n",
      " |  .. [2] https://github.com/piskvorky/gensim/wiki/Recipes-&-FAQ#q4-how-do-you-output-the-u-s-vt-matrices-of-lsi\n",
      " |  \n",
      " |  Method resolution order:\n",
      " |      LsiModel\n",
      " |      gensim.interfaces.TransformationABC\n",
      " |      gensim.utils.SaveLoad\n",
      " |      gensim.models.basemodel.BaseTopicModel\n",
      " |      builtins.object\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __getitem__(self, bow, scaled=False, chunksize=512)\n",
      " |      Return latent representation, as a list of (topic_id, topic_value) 2-tuples.\n",
      " |      \n",
      " |      This is done by folding input document into the latent topic space.\n",
      " |      \n",
      " |      If `scaled` is set, scale topics by the inverse of singular values (default: no scaling).\n",
      " |  \n",
      " |  __init__(self, corpus=None, num_topics=200, id2word=None, chunksize=20000, decay=1.0, distributed=False, onepass=True, power_iters=2, extra_samples=100)\n",
      " |      `num_topics` is the number of requested factors (latent dimensions).\n",
      " |      \n",
      " |      After the model has been trained, you can estimate topics for an\n",
      " |      arbitrary, unseen document, using the ``topics = self[document]`` dictionary\n",
      " |      notation. You can also add new training documents, with ``self.add_documents``,\n",
      " |      so that training can be stopped and resumed at any time, and the\n",
      " |      LSI transformation is available at any point.\n",
      " |      \n",
      " |      If you specify a `corpus`, it will be used to train the model. See the\n",
      " |      method `add_documents` for a description of the `chunksize` and `decay` parameters.\n",
      " |      \n",
      " |      Turn `onepass` off to force a multi-pass stochastic algorithm.\n",
      " |      \n",
      " |      `power_iters` and `extra_samples` affect the accuracy of the stochastic\n",
      " |      multi-pass algorithm, which is used either internally (`onepass=True`) or\n",
      " |      as the front-end algorithm (`onepass=False`). Increasing the number of\n",
      " |      power iterations improves accuracy, but lowers performance. See [3]_ for\n",
      " |      some hard numbers.\n",
      " |      \n",
      " |      Turn on `distributed` to enable distributed computing.\n",
      " |      \n",
      " |      Example:\n",
      " |      \n",
      " |      >>> lsi = LsiModel(corpus, num_topics=10)\n",
      " |      >>> print(lsi[doc_tfidf]) # project some document into LSI space\n",
      " |      >>> lsi.add_documents(corpus2) # update LSI on additional documents\n",
      " |      >>> print(lsi[doc_tfidf])\n",
      " |      \n",
      " |      .. [3] http://nlp.fi.muni.cz/~xrehurek/nips/rehurek_nips.pdf\n",
      " |  \n",
      " |  __str__(self)\n",
      " |      Return str(self).\n",
      " |  \n",
      " |  add_documents(self, corpus, chunksize=None, decay=None)\n",
      " |      Update singular value decomposition to take into account a new\n",
      " |      corpus of documents.\n",
      " |      \n",
      " |      Training proceeds in chunks of `chunksize` documents at a time. The size of\n",
      " |      `chunksize` is a tradeoff between increased speed (bigger `chunksize`)\n",
      " |      vs. lower memory footprint (smaller `chunksize`). If the distributed mode\n",
      " |      is on, each chunk is sent to a different worker/computer.\n",
      " |      \n",
      " |      Setting `decay` < 1.0 causes re-orientation towards new data trends in the\n",
      " |      input document stream, by giving less emphasis to old observations. This allows\n",
      " |      LSA to gradually \"forget\" old observations (documents) and give more\n",
      " |      preference to new ones.\n",
      " |  \n",
      " |  print_debug(self, num_topics=5, num_words=10)\n",
      " |      Print (to log) the most salient words of the first `num_topics` topics.\n",
      " |      \n",
      " |      Unlike `print_topics()`, this looks for words that are significant for a\n",
      " |      particular topic *and* not for others. This *should* result in a more\n",
      " |      human-interpretable description of topics.\n",
      " |  \n",
      " |  save(self, fname, *args, **kwargs)\n",
      " |      Save the model to file.\n",
      " |      \n",
      " |      Large internal arrays may be stored into separate files, with `fname` as prefix.\n",
      " |      \n",
      " |      Note: do not save as a compressed file if you intend to load the file back with `mmap`.\n",
      " |  \n",
      " |  show_topic(self, topicno, topn=10)\n",
      " |      Return a specified topic (=left singular vector), 0 <= `topicno` < `self.num_topics`,\n",
      " |      as a string.\n",
      " |      \n",
      " |      Return only the `topn` words which contribute the most to the direction\n",
      " |      of the topic (both negative and positive).\n",
      " |      \n",
      " |      >>> lsimodel.show_topic(10, topn=5)\n",
      " |      [(\"category\", -0.340), (\"$M$\", 0.298), (\"algebra\", 0.183), (\"functor\", -0.174), (\"operator\", -0.168)]\n",
      " |  \n",
      " |  show_topics(self, num_topics=-1, num_words=10, log=False, formatted=True)\n",
      " |      Return `num_topics` most significant topics (return all by default).\n",
      " |      For each topic, show `num_words` most significant words (10 words by default).\n",
      " |      \n",
      " |      The topics are returned as a list -- a list of strings if `formatted` is\n",
      " |      True, or a list of `(word, probability)` 2-tuples if False.\n",
      " |      \n",
      " |      If `log` is True, also output this result to log.\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Class methods defined here:\n",
      " |  \n",
      " |  load(fname, *args, **kwargs) from builtins.type\n",
      " |      Load a previously saved object from file (also see `save`).\n",
      " |      \n",
      " |      Large arrays can be memmap'ed back as read-only (shared memory) by setting `mmap='r'`:\n",
      " |      \n",
      " |          >>> LsiModel.load(fname, mmap='r')\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors inherited from gensim.utils.SaveLoad:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Methods inherited from gensim.models.basemodel.BaseTopicModel:\n",
      " |  \n",
      " |  print_topic(self, topicno, topn=10)\n",
      " |      Return a single topic as a formatted string. See `show_topic()` for parameters.\n",
      " |      \n",
      " |      >>> lsimodel.print_topic(10, topn=5)\n",
      " |      '-0.340 * \"category\" + 0.298 * \"$M$\" + 0.183 * \"algebra\" + -0.174 * \"functor\" + -0.168 * \"operator\"'\n",
      " |  \n",
      " |  print_topics(self, num_topics=20, num_words=10)\n",
      " |      Alias for `show_topics()` that prints the `num_words` most\n",
      " |      probable words for `topics` number of topics to log.\n",
      " |      Set `topics=-1` to print all topics.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(gensim.models.LsiModel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "corpus_lsi2 = lsi2[tfidf]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[(0, -0.013810624651677433), (1, -0.014381380845581099), (2, -0.023466855270434073), (3, -0.0032649395288155804), (4, 0.026375481836280371), (5, 0.0097529015772247433), (6, -0.0041502781020606762), (7, -0.0036633208709689231), (8, 0.024745103847704782), (9, 0.0032320827214444635)]\n",
      "[(0, -0.018763426107994727), (1, -0.021670110775621341), (2, -0.086748705074664631), (3, -0.0022965198652322239), (4, -0.019916594462958966), (5, -0.0010333695079973301), (6, -0.010291101930211049), (7, -0.0033087173866814122), (8, 0.018762380075057643), (9, 0.00092582382661085148)]\n",
      "[(0, -0.018763426107994727), (1, -0.021670110775621341), (2, -0.086748705074664631), (3, -0.0022965198652322239), (4, -0.019916594462958966), (5, -0.0010333695079973301), (6, -0.010291101930211049), (7, -0.0033087173866814122), (8, 0.018762380075057643), (9, 0.00092582382661085148)]\n",
      "[(0, -0.02626840437841434), (1, -0.019399086741583669), (2, -0.057293268413171498), (3, -0.0015597029170754328), (4, 0.036643312616081093), (5, 0.01797501579379341), (6, 0.00092164553562497982), (7, 0.0014356653894482734), (8, 0.027183563694062579), (9, 0.0029005377949399423)]\n",
      "[(0, -0.041422437237074547), (1, -0.02574927598800953), (2, -0.083846837955137477), (3, -0.018926603604427071), (4, 0.0007431641842075697), (5, 0.0060662253296171207), (6, -0.012478947825271572), (7, -0.0069070267725799932), (8, 0.038100835743295675), (9, 0.0045274422419267196)]\n",
      "[(0, -0.041422437237074547), (1, -0.02574927598800953), (2, -0.083846837955137477), (3, -0.018926603604427071), (4, 0.0007431641842075697), (5, 0.0060662253296171207), (6, -0.012478947825271572), (7, -0.0069070267725799932), (8, 0.038100835743295675), (9, 0.0045274422419267196)]\n",
      "[(0, -0.041422437237074547), (1, -0.02574927598800953), (2, -0.083846837955137477), (3, -0.018926603604427071), (4, 0.0007431641842075697), (5, 0.0060662253296171207), (6, -0.012478947825271572), (7, -0.0069070267725799932), (8, 0.038100835743295675), (9, 0.0045274422419267196)]\n",
      "[(0, -0.030631310037429874), (1, -0.028777607884008873), (2, -0.082343858196554584), (3, -0.0071793517716058413), (4, 0.056764717102693277), (5, 0.037444487917295551), (6, 0.0077447389871331679), (7, 0.0070593210972088958), (8, 0.035578507394282859), (9, 0.0078436414692346149)]\n",
      "[(0, -0.026142370788652079), (1, -0.021310502761664032), (2, -0.059856092422650026), (3, -0.012589427952327183), (4, 0.019371144258552086), (5, 0.01348997606782271), (6, -0.006214731883300904), (7, -0.0041613084461190842), (8, 0.034643845068959365), (9, 0.0061626832496696135)]\n",
      "[(0, -0.012764151182494386), (1, -0.014412079412326542), (2, -0.034720487912886747), (3, -0.0066573088979915351), (4, 0.024427860072552206), (5, 0.014238097357013329), (6, -0.0017189760677917968), (7, -0.0018751083174406698), (8, 0.025308107917107234), (9, 0.0056266733755283441)]\n"
     ]
    }
   ],
   "source": [
    "for i, doc in enumerate(corpus_lsi2):\n",
    "    print(doc)\n",
    "    if i >= 9: break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3321"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(corpus_lsi2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X = gensim.matutils.corpus2dense(corpus_lsi2, num_terms=10, num_docs=len(corpus_lsi2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(10, 3321)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-0.01381062, -0.01876343, -0.01876343, ..., -0.00933241,\n",
       "        -0.02392174, -0.02397908],\n",
       "       [-0.01438138, -0.02167011, -0.02167011, ..., -0.00928216,\n",
       "        -0.01902175, -0.01797347],\n",
       "       [-0.02346686, -0.0867487 , -0.0867487 , ..., -0.01932537,\n",
       "        -0.0404932 , -0.03679732],\n",
       "       ..., \n",
       "       [-0.00366332, -0.00330872, -0.00330872, ..., -0.00203307,\n",
       "        -0.00257692, -0.00132331],\n",
       "       [ 0.0247451 ,  0.01876238,  0.01876238, ...,  0.01016099,\n",
       "         0.03106793,  0.02933426],\n",
       "       [ 0.00323208,  0.00092582,  0.00092582, ..., -0.00075495,\n",
       "        -0.00358354, -0.00377633]], dtype=float32)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "X=X.T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[-0.01381062, -0.01438138, -0.02346686, ..., -0.00366332,\n",
       "         0.0247451 ,  0.00323208],\n",
       "       [-0.01876343, -0.02167011, -0.0867487 , ..., -0.00330872,\n",
       "         0.01876238,  0.00092582],\n",
       "       [-0.01876343, -0.02167011, -0.0867487 , ..., -0.00330872,\n",
       "         0.01876238,  0.00092582],\n",
       "       ..., \n",
       "       [-0.00933241, -0.00928216, -0.01932537, ..., -0.00203307,\n",
       "         0.01016099, -0.00075495],\n",
       "       [-0.02392174, -0.01902175, -0.0404932 , ..., -0.00257692,\n",
       "         0.03106793, -0.00358354],\n",
       "       [-0.02397908, -0.01797347, -0.03679732, ..., -0.00132331,\n",
       "         0.02933426, -0.00377633]], dtype=float32)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [conda root]",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
